Enabling AI chat flag adds the Brave AI button in between side bar items

vadimstruts opened the pull request:
Fixed Add/Move/Remove behavior for the Built-In icon types on the sidebar for the next cases:
- Just after enabling the Ai-Chat, i.e. the AI-Chat icon (or any other built-in) has not been ever seen (or never persisted)
- when add button clicked, add the Built-In icon once after last visible built-in icons on the sidebar

More behavior details in the `Test Plan` below

<!-- Add brave-browser issue below that this PR will resolve -->
Resolves https://github.com/brave/brave-browser/issues/30256

<!-- CI-related labels that can be applied to this PR:
* CI/run-audit-deps (1) - check for known npm/cargo vulnerabilities (audit_deps)
* CI/run-network-audit (1) - run network-audit
* CI/run-upstream-tests - run Chromium unit and browser tests on Linux and Windows (otherwise only on Linux)
* CI/skip - do not run CI builds (except noplatform)
* CI/skip-linux-arm64, CI/skip-linux-x64, CI/skip-android, CI/skip-macos, CI/skip-ios, CI/skip-windows-arm64, CI/skip-windows-x64, CI/skip-windows-x86 - skip CI builds for specific platforms
* CI/skip-upstream-tests - do not run Chromium unit, or browser tests (otherwise only on Linux)
* CI/skip-all-linters - do not run presubmit and lint checks
* CI/storybook-url (1) - deploy storybook and provide a unique URL for each build

(1) applied automatically when some files are changed (see: https://github.com/brave/brave-core/blob/master/.github/labeler.yml)
-->

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
1.  Start the browser first time, just after enabling the Ai-Chat. 
    It means AI-Chat icon has not been ever seen (or never persisted)
    
    **Expected result:** AI-Chat item appears just behind the last built-in icon (i.e. Reading List)
    
2.  Move the AI-Chat icon up or down to make sure that its position has been added to 
    the prefs.
    Restart the browser.
    
    **Expected result:** The AI-Chat icon appears on the same place.

3.  Remove the AI-Chat icon and then add it again
    
    **Expected result:** The AI-Chat icon added just after the last built-in icon (i.e. Reading List)

simonhong commented:
UI has assumption that newly added item is always added at last.
Need to update UI side also to remove that assumption. I'l working on it.

petemill commented on components/sidebar/sidebar_service.cc:
```diff
@@ -72,6 +74,13 @@ SidebarItem::BuiltInItemType GetBuiltInItemTypeForLegacyURL(
+  const auto it =
+      std::find_if(items.rbegin(), items.rend(),
+                   [](const SidebarItem& item) { return IsBuiltInType(item); });
+  return items.rend() - it;
```
Is this going to end up being index `0` if no built-in items are found? I think that's what we want.

vadimstruts replied:
Yes, if no built-in items, the result is `0`

petemill reviewed (changes requested):
As far as I'm understanding this, I don't think it's following the request in the issue at https://github.com/brave/brave-browser/issues/30256#issuecomment-1611791362

> Move AI chat button to below Wallet by default.

I appreciate this PR attempts to solve for both users who have not customized Sidebar and users who have, but I think that's mostly catered for in the current `LoadSidebarItems` implementation. And simply changing the order in `sidebar_service.h`'s `kDefaultBuiltInItemTypes` would have likely satisfied the issue request for the most part.

However, if we do want to satisfy users who have re-ordered and added custom items before the built-in items, perhaps we can search for each built-in item preceding each new built-in item and insert just after it? For example, if the intended default order is Talk, Wallet, AIChat, and AIChat is the new item in LoadSidebarItems, Wallet should be before the new AIChat item, so wherever Wallet is, insert AIChat after. And if Wallet is hidden by the user, find Talk and insert after that. And if that has also been removed, then insert the new item at index `0`)?

I like your addition to `SidebarService::AddItem`, although it doesn't seem too important as the user must first remove the built-in item and then re-add, so adding at the end seems intuitive. But if we do want to try the change, perhaps in `AddItem` we can use the same alogithm I just described for `LoadSidebarItems`?

petemill commented on components/sidebar/sidebar_service.cc:
```diff
@@ -533,25 +549,13 @@ void SidebarService::LoadSidebarItems() {
-    // the middle of custom items, but that will still work.
-    auto index = std::min(static_cast<int>(default_index),
-                          static_cast<int>(items_.size()));
+    const size_t pos_to_insert = GetLastBuiltInItemIndex(items());
```
Yeah I'm afraid I don't think this is what we want to do in this function as this would mean any new item we add would always be at the end for existing users. And we want to insert new items at a predictable index, especially for the majority users who haven't customized sidebar items.

vadimstruts replied:
removed that

simonhong commented:
Added https://github.com/brave/brave-core/pull/19637/commits/f319231d426f9c984af6a36bb62a78b6961c6e70.
So far we anchored  the `item added bubble`  to last item as it's inserted at last always.
Also,  we scrolled to bottom when it's scrollable area after inserting new item. With this change, scroll is done to make new item visible regardless of its inserted index.

vadimstruts commented:
Added fix, based on the next described logic:
> However, if we do want to satisfy users who have re-ordered and added custom items before the built-in items, perhaps we can search for each built-in item preceding each new built-in item and insert just after it? For example, if the intended default order is Talk, Wallet, AIChat, and AIChat is the new item in LoadSidebarItems, Wallet should be before the new AIChat item, so wherever Wallet is, insert AIChat after. And if Wallet is hidden by the user, find Talk and insert after that. And if that has also been removed, then insert the new item at index `0`)?

It is added to the `SidebarService::AddItem`, like you said here:
> I like your addition to `SidebarService::AddItem`, although it doesn't seem too important as the user must first remove the built-in item and then re-add, so adding at the end seems intuitive. But if we do want to try the change, perhaps in `AddItem` we can use the same alogithm I just described for `LoadSidebarItems`?

Here is video how it works: https://github.com/brave/brave-core/assets/118171981/76aae3fe-13b5-49fb-a686-b57e286040d4

petemill commented on components/sidebar/sidebar_service.cc:
```diff
@@ -72,6 +72,53 @@ SidebarItem::BuiltInItemType GetBuiltInItemTypeForLegacyURL(
+  auto* iter = std::find(iter_begin, iter_end, item);
+
+  if (iter_end == iter) {
+    return *std::prev(iter_end);
```
`DCHECK` that the previous item is not `iter_begin`, i.e. there are items in `kDefaultBuiltInItemTypes`? Or just let this fall through to the logic below?

simonhong replied:
How about adding this instead of this if clause?
```
CHECK(iter_end != iter);
```
With current constraint, `kDefaultBuiltInItemTypes` should include all.
```
      std::size(kDefaultBuiltInItemTypes) ==
          static_cast<size_t>(SidebarItem::BuiltInItemType::kBuiltInItemLast),
```

petemill commented on components/sidebar/sidebar_service_unittest.cc:
```diff
@@ -101,7 +240,7 @@ TEST_F(SidebarServiceTest, AddRemoveItems) {
 
   // Add again.
   // New item will be added at last.
```
maybe the comment needs changing as it won't be the last position now

petemill commented on components/sidebar/sidebar_service_unittest.cc:
```diff
@@ -729,4 +877,175 @@ TEST_F(SidebarServiceTestWithPlaylist, GetDefaultPanelItem) {
+    scoped_feature_list_.InitAndEnableFeature(ai_chat::features::kAIChat);
+#endif  // BUILDFLAG(ENABLE_AI_CHAT)
+
+    auto registry = base::MakeRefCounted<user_prefs::PrefRegistrySyncable>();
+    sidebar::SidebarServiceFactory::GetInstance();
+    RegisterUserProfilePrefs(registry.get());
+    registry->RegisterDictionaryPref("brave");
+    registry->RegisterDictionaryPref("brave.sidebar");
+
+    sync_preferences::PrefServiceMockFactory factory;
+    auto pref_service = factory.CreateSyncable(registry.get());
+    auto* pref_service_ptr = pref_service.get();
+
+    auto builder = TestingProfile::Builder();
+    builder.SetPrefService(std::move(pref_service));
+    profile_ = builder.Build();
+
+    ASSERT_EQ(pref_service_ptr, profile_->GetPrefs());
```
What's the reason for doing this instead of having a simpler `TestingPrefServiceSimple` and calling SidebarService::RegisterProfilePrefs`? Perhaps add a comment here as to why it's this way.

simonhong replied:
I think you could subclass `SidebarServiceTest` and add additional code for testing ai chat
as `SidebarServiceTest` already has similar `InitService()`.

vadimstruts replied:
done

petemill commented on components/sidebar/sidebar_service_unittest.cc:
```diff
@@ -729,4 +877,175 @@ TEST_F(SidebarServiceTestWithPlaylist, GetDefaultPanelItem) {
+                        SidebarItem::BuiltInItemType::kBookmarks,
+                        SidebarItem::BuiltInItemType::kBraveTalk,
+                    },
+#if BUILDFLAG(ENABLE_AI_CHAT)
+                    8UL
+#else
+                    7UL
+#endif  // BUILDFLAG(ENABLE_AI_CHAT)
```
There's a lot of manual specifying here, and a lot of hard-coded counts and lists that will need changing when we add or modify sidebar item defaults. But perhaps that's what we want for tests? Would be interested to see @simonhong's opinion.

simonhong replied:
Hmm, I think this is fine as it's based on preset data such as `sidebar_all_builtin_visible_json` and it could be used for checking preset is loaded properly.
If default item list change affects this preset data, both (preset data and this count number) should be updated.
I would suggest that it would be more better to have another constants like `sidebar_all_builtin_visible_json_total_item_count` instead of using `8UL` or `7UL`.

```suggestion
#if BUILDFLAG(ENABLE_AI_CHAT)
                    sidebar_all_builtin_visible_json_total_item_count
#else
                    sidebar_all_builtin_visible_json_total_item_count - 1
#endif  // BUILDFLAG(ENABLE_AI_CHAT)
```

vadimstruts replied:
done

petemill reviewed (approved):
Thanks, this is great. I think it works well, but we'll have to see how it feels when we're using it. Please wait for @simonhong review if possible.

simonhong commented on components/sidebar/sidebar_service_unittest.cc:
```diff
@@ -729,4 +877,175 @@ TEST_F(SidebarServiceTestWithPlaylist, GetDefaultPanelItem) {
+  }
+
+  void LoadFromPrefsTest(
+      base::Value::Dict sidebar_prefs,
+      const std::vector<SidebarItem::BuiltInItemType>& defined_order,
+      const size_t& expected_items_loaded) {
```
nit:
```suggestion
      const base::Value::Dict& sidebar_prefs,
      const std::vector<SidebarItem::BuiltInItemType>& defined_order,
      const size_t expected_items_loaded) {
```

vadimstruts replied:
done

simonhong commented on components/sidebar/sidebar_service_unittest.cc:
```diff
@@ -729,4 +877,175 @@ TEST_F(SidebarServiceTestWithPlaylist, GetDefaultPanelItem) {
+      base::Value::Dict sidebar_prefs,
+      const std::vector<SidebarItem::BuiltInItemType>& defined_order,
+      const size_t& expected_items_loaded) {
+    // base::Value::Dict
+    // sidebar(ParseTestJson(sidebar_all_builtin_visible_json));
```
left over?

vadimstruts replied:
yes, removed it

simonhong commented on components/sidebar/sidebar_service_unittest.cc:
```diff
@@ -729,4 +877,175 @@ TEST_F(SidebarServiceTestWithPlaylist, GetDefaultPanelItem) {
+    // sidebar(ParseTestJson(sidebar_all_builtin_visible_json));
+    GetPrefs()->Set(kSidebarItems,
+                    std::move(sidebar_prefs.Find("sidebar_items")->Clone()));
+    GetPrefs()->SetInteger(kSidebarShowOption, 0);
+    EXPECT_EQ(0, GetPrefs()->GetInteger(kSidebarShowOption));
```
Maybe you want to set it from `sidebar_prefs` as you defined it in preset data?
but I think setting/checking show option here doesn't have much meaning.

vadimstruts replied:
removed it

simonhong commented on components/sidebar/sidebar_service_unittest.cc:
```diff
@@ -729,4 +880,152 @@ TEST_F(SidebarServiceTestWithPlaylist, GetDefaultPanelItem) {
+  ~SidebarServiceOrderingTest() override = default;
+
+  void SetUp() override {
+    testing::Test::SetUp();
```
```suggestion
    SidebarServiceTest::SetUp();
```

vadimstruts replied:
removed it

simonhong reviewed (approved):
++ 👍🏼

sangwoo108 commented on components/sidebar/sidebar_service_unittest.cc:
```diff
@@ -729,4 +880,151 @@ TEST_F(SidebarServiceTestWithPlaylist, GetDefaultPanelItem) {
+                    sidebar_items->size() - 1
+#endif  // BUILDFLAG(ENABLE_AI_CHAT)
+  );
+}
+TEST_F(SidebarServiceOrderingTest, LoadFromPrefsWalletBuiltInHidden) {
```
nit: 
```suggestion
}

TEST_F(SidebarServiceOrderingTest, LoadFromPrefsWalletBuiltInHidden) {
```

vadimstruts replied:
fixed

sangwoo108 commented on components/sidebar/sidebar_service_unittest.cc:
```diff
@@ -729,4 +880,151 @@ TEST_F(SidebarServiceTestWithPlaylist, GetDefaultPanelItem) {
+                    sidebar_items->size() - 1
+#endif  // BUILDFLAG(ENABLE_AI_CHAT)
+  );
+}
+TEST_F(SidebarServiceOrderingTest, LoadFromPrefsAiChatBuiltInNotListed) {
+  base::Value::Dict sidebar(
```
nit:

```suggestion
}

TEST_F(SidebarServiceOrderingTest, LoadFromPrefsAiChatBuiltInNotListed) {
  base::Value::Dict sidebar(
```

vadimstruts replied:
fixed

sangwoo108 reviewed (approved):
LGTM 👍
